File: src/utils/runtime-actions.ts
import type { RuntimeAction } from "../types/cycle";

export interface RuntimeMarkerOption {
  id: string;
  name: string;
  allowed_values?: string[];
}

export interface RuntimeTimerOption {
  id: string;
  name: string;
}

export interface RuntimeCounterOption {
  id: string;
  name: string;
}

export type RuntimeActionType = RuntimeAction["type"];

export function markerValueOptions(
  markers: RuntimeMarkerOption[] | undefined,
  markerId: string
): string[] {
  return markers?.find((marker) => marker.id === markerId)?.allowed_values ?? [];
}

export function createDefaultRuntimeAction(
  type: RuntimeActionType,
  markers: RuntimeMarkerOption[] | undefined,
  timers: RuntimeTimerOption[] | undefined,
  counters: RuntimeCounterOption[] | undefined = []
): RuntimeAction {
  const marker = markers?.[0];
  const timer = timers?.[0];
  const counter = counters?.[0];

  switch (type) {
    case "set_marker":
      return {
        type,
        marker_id: marker?.id ?? "",
        value: marker?.allowed_values?.[0] ?? "",
      };
    case "clear_marker":
      return {
        type,
        marker_id: marker?.id ?? "",
      };
    case "record_timer":
    case "reset_timer":
      return {
        type,
        timer_id: timer?.id ?? "",
      };
    case "increment_counter":
      return {
        type,
        counter_id: counter?.id ?? "",
        by: 1,
      };
    case "set_counter":
      return {
        type,
        counter_id: counter?.id ?? "",
        value: 0,
      };
    case "reset_counter":
      return {
        type,
        counter_id: counter?.id ?? "",
      };
  }
}

export function runtimeActionTitle(type: RuntimeActionType): string {
  switch (type) {
    case "set_marker":
      return "设置标记";
    case "clear_marker":
      return "清除标记";
    case "record_timer":
      return "记录时间";
    case "reset_timer":
      return "重置时间";
    case "increment_counter":
      return "增加计数";
    case "set_counter":
      return "设置计数";
    case "reset_counter":
      return "重置计数";
  }
}

File: src-tauri/src/engine/runtime_actions.rs
//! Runtime action application for `CycleExecutor`.

use crate::engine::cycle_executor::{CycleExecutor, CycleLogEvent};
use crate::models::cycle::RuntimeAction;

impl<'a> CycleExecutor<'a> {
    pub(super) fn apply_runtime_actions(
        &mut self,
        actions: &[RuntimeAction],
        now_ms: u64,
        phase_idx: usize,
        phase_name: &str,
        skill_id: &str,
    ) {
        for action in actions {
            let reason = match action {
                RuntimeAction::SetMarker { marker_id, value } => {
                    self.runtime.set_marker(marker_id, value);
                    format!("set_marker:{marker_id}={value}")
                }
                RuntimeAction::ClearMarker { marker_id } => {
                    self.runtime.clear_marker(marker_id);
                    format!("clear_marker:{marker_id}")
                }
                RuntimeAction::RecordTimer { timer_id } => {
                    self.runtime.record_timer(timer_id, now_ms);
                    format!("record_timer:{timer_id}")
                }
                RuntimeAction::ResetTimer { timer_id } => {
                    self.runtime.reset_timer(timer_id);
                    format!("reset_timer:{timer_id}")
                }
                RuntimeAction::IncrementCounter { counter_id, by } => {
                    self.runtime.increment_counter(counter_id, *by);
                    format!("increment_counter:{counter_id}+={by}")
                }
                RuntimeAction::SetCounter { counter_id, value } => {
                    self.runtime.set_counter(counter_id, *value);
                    format!("set_counter:{counter_id}={value}")
                }
                RuntimeAction::ResetCounter { counter_id } => {
                    self.reset_counter_to_initial(counter_id);
                    format!("reset_counter:{counter_id}")
                }
            };
            self.log_event(CycleLogEvent {
                ts_ms: now_ms,
                phase_index: phase_idx,
                phase_name,
                event: "runtime_action",
                skill_id,
                outcome: "Applied",
                reason: &reason,
            });
        }
    }

    pub(super) fn reset_phase_entry_counters(&mut self) {
        let Some(schema) = &self.config.state_schema else {
            return;
        };
        for counter in &schema.counters {
            if counter.reset_on_phase_entry {
                self.runtime.set_counter(&counter.id, counter.initial_value);
            }
        }
    }

    fn reset_counter_to_initial(&mut self, counter_id: &str) {
        let counter_id = counter_id.trim();
        if counter_id.is_empty() {
            return;
        }
        let initial_value = self
            .config
            .state_schema
            .as_ref()
            .and_then(|schema| {
                schema
                    .counters
                    .iter()
                    .find(|counter| counter.id == counter_id)
            })
            .map(|counter| counter.initial_value)
            .unwrap_or(0);
        self.runtime.set_counter(counter_id, initial_value);
    }
}
